docs/pallas/tpu/distributed.md

```diff
@@ -0,0 +1,1453 @@
+
++++ {"id": "1Oc_WD1hChfN"}
+
+## Remote Direct Memory Access (DMA) Model
+
+TPUs communicate via the remote DMA model, which is a restricted form of communication between devices. In traditional multi-threaded programming, different threads can typically read and write to arbitrary locations in shared memory. On TPUs, any buffers allocated to a kernel (in SMEM, VMEM, or HBM) are visible to all devices, and communication happens only in terms of writes from one device's local buffer to another device's buffer via a combined send-receive operation.
+
+### Async Remote Copy Operation
+The `make_async_remote_copy(src_ref, dst_ref, send_sem, recv_sem, device_id, device_id_type)` function is used to initiate a remote DMA for simultaneously sending data to one device while receiving data from another device. Let's break down what each argument means:
+
+- `src_ref` is the local `Ref` (in any memory space) containing the data you wish to send to `dst_ref` on another device.
+- `dst_ref` is the remote `Ref` (in any memory space) at which data will be copied to on the target device.
+- `send_sem` is a DMA semaphore used to block until all data has been sent from `src_ref`.
+- `recv_sem` is a DMA semaphore used to block until the expected number of bytes have been received at `dst_ref`. The sender of the DMA will write to the receiver's `recv_sem`.
+- `device_id` is the device ID of the target device to send to.
+- `device_id_type` specifies the format of `device_id`, which can either be in LOGICAL format (integer device ID), or in MESH format (an ND-tuple index into the logical device mesh). The default mode is MESH.
+
+`make_async_remote_copy` returns a descriptor object on which you use the `.start()` method to initiate the DMA, and the `.wait_send()` to block on `send_sem` and `.wait_recv()` to block on `recv_sem` (or `.wait()` to block on both). If a device is only expected to send data, it is sufficient to only call `.start()` and `.wait_send()`, and likewise if a device is only receiving it is sufficient to only call `.wait_recv()`. If using a SIMD pattern where all devices execute the DMA, each device will generally call both `.start()` and `.wait()`.
+```python
+dma_descriptor = make_async_remote_copy(src_ref, dst_ref, send_sem, recv_sem, device_id)
+dma_descriptor.start() # Initiate the DMA (non-blocking).
+# ... do other work
+dma_descriptor.wait_send() # Block until all data has been sent.
+dma_descriptor.wait_recv() # Block until all data has been received.
+```
+
+As an example, let's visualize a DMA where we consider 4 devices (indexed 0, 1, 2, 3). We consider a scheme where device 0 copies to device 1, and device 2 & 3 copy to each other. In practice, we can create such an asymmetric communication pattern by using `@pl.when` to branch on the device ID.
+
+(1) Each device creates the DMA descriptor. Devices 0, 2, and 3 call `.start()` to initiate the DMA from `src_ref`. Device 1 is idle.
```

Comment:
```suggestion
(1) Each device creates the DMA descriptor. Devices 0, 2, and 3 call `.start()` to initiate the DMA from `src_ref`. Device 1 skips the `.start()`, e.g. by using `pl.when`.
```

Even though it's mentioned in the sentence above, I like saying it here for concreteness in the first step.

Reply:
Done.